tests/5.0/target/test_target_device.F90

```diff
@@ -0,0 +1,99 @@
+!===--- test_target_device.F90 -------------------------------===//
+!
+! OpenMP API Version 5.0 Nov 2018
+!
+! This test checks the target construct with device clause where device-
+! modifier is either ancestor or device_num. If no device_modifier is 
+! present, the behavior is the same as if device_num were present.
+! 
+!//===----------------------------------------------------------------------===//
+#define OMPVV_MODULE_REQUIRES_LINE !$omp requires reverse_offload
+#include "ompvv.F90"
+
+#define N 1024
+
+PROGRAM test_target_device
+  ! Required for device(ancestor: 1)
+  !$omp requires reverse_offload
+  USE iso_fortran_env
+  USE ompvv_lib
+  USE omp_lib
+  implicit none
+
+  OMPVV_TEST_OFFLOADING
+
+  OMPVV_TEST_VERBOSE(target_device_ancestor() .ne. 0)
+  OMPVV_TEST_VERBOSE(target_device_device_num() .ne. 0)
+
+  OMPVV_REPORT_AND_RETURN()
+
+CONTAINS
+  INTEGER FUNCTION target_device_ancestor()
+    INTEGER :: errors, i, which_device
+    INTEGER, DIMENSION(N) :: a
+    LOGICAL :: is_shared_env
+
+    errors = 0
+    which_device = 0
+    is_shared_env = .false.
+
+    DO i = 1, N
+       a(i) = i
+    END DO
+
+    OMPVV_TEST_AND_SET(errors, omp_get_num_devices() .le. 0) 
+    OMPVV_WARNING_IF(omp_get_num_devices() .le. 0, "[SKIPPED] Since no target devices were found, this test will be skipped.")
+    OMPVV_TEST_AND_SET_SHARED_ENVIRONMENT(is_shared_env)
+    OMPVV_WARNING_IF(is_shared_env, "[WARNING] target_device_ancestor() test may not be able to detect errors if the target system supports shared memory.")
+
+
+  !$omp target ! Run on the default device, which is the host for device_num = 0
+    !$omp target device(ancestor: 1) map(tofrom: a) map(to: which_device)
+    DO i = 1, N
+       a(i) = a(i) + 2
+    END DO
+    which_device = 75
+    !$omp end target
+  !$omp end target
```

Comment:
The expected value on the host is 0 not 75:  `!$omp target` has an implicit `firstprivate(which_device)` – and this firstprivatized variable is then mapped to the device explicitly (`map(to: which_device) `).

Without (unified) shared memory, the expected value after the inner/reverse-offload region is `0` (not copied back to the device) – and (with and without shared memory) on outside of all target regions, it should be 0.

I think you need something like `map(to: which_device)` on the outer `target` directive.

Additionally, given that 'a' was set, shouldn't there be some check like:
```F90
    OMPVV_TEST_AND_SET(errors, any (a /= [(i+2, i =1, N)]))
```
albeit without using `map(to: a)`, there is a `map(tofrom: a)` clause on both directives: on the outer implicitly and on the inner explicitly. Hence, also without shared memory and without working reverse offload, it would fail.

Maybe something like that (as the next line mentions, 'errors2' still needs to be declared. There is no ..._VERBOSE inside the target region as not all implementations support I/O in the target region):
```suggestion
  errors2 = 0  ! a second variable - needs to be declared
  !$omp target map(tofrom: errors2) map(to:a, which_device) ! Run on the default device, which is the host for device_num = 0
    !$omp target device(ancestor: 1) map(tofrom: a) map(to: which_device)
    DO i = 1, N
       a(i) = a(i) + 2
    END DO
    which_device = 75
    !$omp end target
    if (which_device /= 75) errors2 = errors2 + 1
    if (any (a /= [(i, i = 1, N)])) errors2 = errors2 + 1
  !$omp end target
  OMPVV_TEST_AND_SET_VERBOSE(errors, errors2 /= 0)
  OMPVV_TEST_AND_SET_VERBOSE(errors, any (a /= [(i+2, i = 1, N)]))
```

Reply:
Updated.

Reply:
Changed `map(tofrom: a)` to `map(to:a)`, and modified the intermediate error checking codes (after inner target region but before the end of the outer target region).